### PR TITLE
Convert quote to order in same transaction

### PR DESCRIPTION
**Pull request by crudmachine**
https://github.com/barbinbrad/carbon/pull/119#discussion_r1556682937 - same transaction for convert quote to order

WIP

### CODE REVIEW

File: apps/carbon/app/routes/x+/quote+/$id.convert-to-order.tsx [hunk 1]
```diff
@@ -18,110 +19,35 @@ export async function action(args: ActionFunctionArgs) {
...
-    });
-
-    if (createSalesOrder.error || !createSalesOrder.data?.[0]) {
-      // TODO: this should be done as a transaction
-      await rollbackNextSequence(client, "salesOrder", userId);
+    const result = await createSalesOrderFromQuote(id, userId);
+    if (result.error || !result?.data) {
       throw redirect(
-        path.to.quotes,
+        request.headers.get("Referer") ?? path.to.quotes,
```
Comment (barbinbrad) [inline apps/carbon/app/routes/x+/quote+/$id.convert-to-order.tsx:35]: There's a util method for this:

```suggestion
        requestReferrer(request) ?? path.to.quote,
```

**Comment by barbinbrad** [comment]
This is really impressive for (I'm guessing) no type checking.  Here is the main problem:

Screenshot 2024-04-29 at 10 19 07 PM

In a kysely transaction, you don't need to check whether each operation succeeded. If it doesnt' succeed it will throw an error, and revert the whole transaction.

To get error checking, I believe you need to do three things:

1. Install Deno
2. Install the Deno VSCode extension
3. cmd+shift+p > Deno: Restart Language Server

**Comment by barbinbrad** [comment]
I think the Convert button here should be disabled unless the status is one of a limited set (maybe Open or Replied). 

Screenshot 2024-04-29 at 10 30 30 PM

**Comment by barbinbrad** [comment]
Doesn't seem like the quantity and unit prices are being carried over for me.


https://github.com/barbinbrad/carbon/assets/64510427/e2928de2-aa0e-4b7a-9b38-0f2715a6650f

File: apps/carbon/app/routes/x+/quote+/$id.convert-to-order.tsx (hunk 1 above)
Comment (crudmachine) [reply apps/carbon/app/routes/x+/quote+/$id.convert-to-order.tsx:35]: addressed and made it `path.to.quote(id)`

**Comment by crudmachine** [comment]
> This is really impressive for (I'm guessing) no type checking. Here is the main problem:
> 
> Screenshot 2024-04-29 at 10 19 07 PM
> 
> In a kysely transaction, you don't need to check whether each operation succeeded. If it doesnt' succeed it will throw an error, and revert the whole transaction.
> 
> To get error checking, I believe you need to do three things:
> 
> 1. Install Deno
> 2. Install the Deno VSCode extension
> 3. cmd+shift+p > Deno: Restart Language Server

Nice! I installed demo and the extension and it gives me some type errors but not the ones in the screenshot - 
Screenshot 2024-05-04 at 6 38 19 PM

Also when I run `deno lint --json` or `deno lint` in the folder of this file, I get no errors in the output. Do you get these errors when you run `deno lint --json` or `deno lint` from the `packages/database/supabase/functions/create-sales-order-from-quote` directory in your terminal?

**Comment by barbinbrad** [comment]
> > This is really impressive for (I'm guessing) no type checking. Here is the main problem:
> > Screenshot 2024-04-29 at 10 19 07 PM
> > In a kysely transaction, you don't need to check whether each operation succeeded. If it doesnt' succeed it will throw an error, and revert the whole transaction.
> > To get error checking, I believe you need to do three things:
> > 
> > 1. Install Deno
> > 2. Install the Deno VSCode extension
> > 3. cmd+shift+p > Deno: Restart Language Server
> 
> Nice! I installed deno and the extension and it gives me some type errors but not the ones in the screenshot - Screenshot 2024-05-04 at 6 38 19 PM
> 
> Also when I run `deno lint --json` or `deno lint` in the folder of this file, I get no errors in the output. Do you get these errors when you run `deno lint --json` or `deno lint` from the `packages/database/supabase/functions/create-sales-order-from-quote` directory in your terminal?

I don't think `deno lint` does what we're looking for:

Screenshot 2024-05-04 at 1 31 55 PM

I'd say if you remove the `if (quoteLinesUpdate.error) throw new Error("Sales order not created");` we should be good to merge. I can fix the rest of it later. Be sure to try cmd + shift + p and the `Deno: Restart Language Server` if you haven't already.

**Pull request closed without merging**